### PR TITLE
TRIB-236: adds ConnectAPI endpoints to get all cosigners for user and all cosigners for a given attribute

### PR DESCRIPTION
**Pull request by mrsbluerose**
Part 1: endpoint to return list of users that cosigned a specified attribute (phrase) for a user
- Creates UsernameDTO with just id and name
- Creates cosign repo method findCosignersByUserIdReceivingAndPhraseId
- Creates cosign service method getCosignersForUserAttribute
- Creates ConnectAPI endpoint getCosignersForUserAttribute

Part 2: endpoint to return a list of all attributes (phrases) cosigned to a user and a list of all users that cosigned each attribute
- Creates cosigns for user DTO
- Creates cosign repo method findAllByUserIdReceiving
- Creates cosign service getAllCosignsForUser
- Creates ConnectAPI endpoint getAllCosignersForUser
- Creates user service getUsernameDTO to build username DTOs

Testing
- Created tests for new methods in cosign service
- Created tests for new method in user service
- Created tests for new endpoints in ConnectAPI
- Ran both endpoints successfully in Postman

### CODE REVIEW

File: src/test/java/com/savvato/tribeapp/services/CosignServiceImplTest.java [hunk 1]
```diff
@@ -91,4 +98,88 @@ public void saveCosignAlreadyExisting() {
...
+        when(userService.getUsernameDTO(anyLong())).thenReturn(mockUsernameDTO);
+
+        // expected results
+        List<UsernameDTO> expectedListUsernameDTO = new ArrayList<>();
+        expectedListUsernameDTO.add(mockUsernameDTO);
+
+        // test
+        List<UsernameDTO> usernameDTOS = cosignService.getCosignersForUserAttribute(testUserIdReceiving,testPhraseId);
+
+        for(UsernameDTO user : usernameDTOS){
```
Comment (haxwell) [inline src/test/java/com/savvato/tribeapp/services/CosignServiceImplTest.java:130]: Create another test, and make the list of expected results three items in length. This test covers the case when there is one element, but not more than one. A test for three items should suffice.

File: src/test/java/com/savvato/tribeapp/services/CosignServiceImplTest.java [hunk 2]
```diff
@@ -91,4 +98,88 @@ public void saveCosignAlreadyExisting() {
...
+            assertEquals(user.userId, expectedListUsernameDTO.get(0).userId);
+            assertEquals(user.username, expectedListUsernameDTO.get(0).username);
+        }
+    }
+
+    @Test
+    public void testGetAllCosignsForUser() {
```
Comment (haxwell) [inline src/test/java/com/savvato/tribeapp/services/CosignServiceImplTest.java:137]: Likewise, create another test, and make the list of expected results three items in length. This test covers the case when there is one element, but not more than one. A test for three items should suffice.

**Comment by haxwell** [comment]
Overall, this is an excellent Pull Request.. Great job! The only thing that should hold it up is the first suggestion, to create a test that uses multiple results when calling the /all endpoint. The other two can be done in another issue/pr if you would like.

**Review by haxwell — CHANGES_REQUESTED** [review]

File: src/test/java/com/savvato/tribeapp/services/CosignServiceImplTest.java (hunk 2 above)
Comment (mrsbluerose) [reply src/test/java/com/savvato/tribeapp/services/CosignServiceImplTest.java:189]: I added a test for three phrases and three users per phrase

File: src/test/java/com/savvato/tribeapp/services/CosignServiceImplTest.java (hunk 1 above)
Comment (mrsbluerose) [reply src/test/java/com/savvato/tribeapp/services/CosignServiceImplTest.java:130]: I added a test for three cosigners

File: src/test/java/com/savvato/tribeapp/controllers/ConnectAPITest.java [hunk 3]
```diff
@@ -370,4 +372,80 @@ public void testGetConnectionsSadPath() throws Exception {
...
+                .perform(
+                        get("/api/connect/cosign/{userIdReceiving}/{phraseId}",testUserIdReceiving,testPhraseId)
+                                .header("Authorization", "Bearer " + auth)
+                                .characterEncoding("utf-8"))
+                .andExpect(status().isOk())
+                .andExpect(content().json("[{\"userId\":1,\"username\":\"test\"}]"));
+    }
+
+    @Test
+    public void testGetAllCosignsForUser() throws Exception {
```
[the remark this answers is not in the document]
Comment (mrsbluerose) [reply src/test/java/com/savvato/tribeapp/controllers/ConnectAPITest.java:411]: I will work on this next week (3/11)

[the remark this answers is not in the document]
Comment (mrsbluerose) [reply src/test/java/com/savvato/tribeapp/controllers/ConnectAPITest.java:411]: Actually, I was able to work on it today. I have it returning data for three phrases, three users each. I also removed the phraseId parameter (mistake)

**Review by haxwell — APPROVED** [review]

**Review by haxwell — CHANGES_REQUESTED** [review]
I noticed one more issue here.. In the log for the CosignServiceImplTest:

```
2024-03-09 09:16:24.769  INFO 1862 --- [           main] c.s.tribeapp.services.CosignServiceImpl  : Cosign from user 1 to user 1 added.
2024-03-09 09:16:24.770  INFO 1862 --- [           main] c.s.tribeapp.services.CosignServiceImpl  : Cosign from user 1 to user 1 added.
2024-03-09 09:16:24.776  INFO 1862 --- [           main] c.s.tribeapp.services.CosignServiceImpl  : Cosign from user 1 to user 1 added.
```

We should not allow a user to cosign themselves. So 
 - the test (or tests) which is producing this should use different user IDs for receiving and issuing
 - and we should have a test that ensures this type of scenario fails.

**Comment by mrsbluerose** [comment]
> I noticed one more issue here.. In the log for the CosignServiceImplTest:
> 
> ```
> 2024-03-09 09:16:24.769  INFO 1862 --- [           main] c.s.tribeapp.services.CosignServiceImpl  : Cosign from user 1 to user 1 added.
> 2024-03-09 09:16:24.770  INFO 1862 --- [           main] c.s.tribeapp.services.CosignServiceImpl  : Cosign from user 1 to user 1 added.
> 2024-03-09 09:16:24.776  INFO 1862 --- [           main] c.s.tribeapp.services.CosignServiceImpl  : Cosign from user 1 to user 1 added.
> ```
> 
> We should not allow a user to cosign themselves. So
> 
> * the test (or tests) which is producing this should use different user IDs for receiving and issuing
> * and we should have a test that ensures this type of scenario fails.

Right... This is due to an error in my logic on a previous ticket with the save cosign methods. As I was working through a quick fix, I felt this was big enough to warrant a new ticket. I am going to write out a ticket and message you on Discord with a link to confer on details. Basically, I think I need to change the cosign service save method to check for duplicates, and the ConnectAPI save cosign endpoint needs to be updated to recognize the cosign was not saved and return a response. My questions for you revolve around what we want to return. I'll write it up and see what you think.

**Review by haxwell — COMMENTED** [review]
lgtm

**Pull request merged**